### PR TITLE
Update LSP Protocol Types

### PR DESCRIPTION
**Pull request by mhutch**
Updates all LSP Protocol types to match the current 3.17 spec. Improves documentation, fixes some minor issues, and adds lots of missing types, properties and methods.

I submoduled Roslyn in the MSBuild Editor repo so I would have definitions to use with CLaSP. After running into some missing things and having to replace some of the definitions locally, I figured the best solution for long-term maintenance would be to update all the definitions to fully cover the spec and contribute them back.

It would be great if these definitions could be included in the CLaSP source package but let's have that discussion separately.

Because of the large amount of changes, I have broken this up into smaller atomic commits structured so that they can be reviewed independently.

I have made a few minor formatting changes and moved a couple files to subdirectories where it made sense together with other changes I was making. These are some bigger changes I'd like to make, such as moving completion types to a `Completion` directory, switching existing code to file scoped namespaces, and changing properties to `required` and `init`, but I'll do those in later PRs if they're acceptable.

Marking as draft as I have not yet checked whether I need to update other Roslyn code to track the handful of API changes I made.

### CODE REVIEW

**Review by dibarbet — COMMENTED** [review]
> These are some bigger changes I'd like to make, such as moving completion types to a Completion directory, switching existing code to file scoped namespaces, and changing properties to required and init, but I'll do those in later PRs if they're acceptable.

I don't remember - is switching to required a binary breaking change?  I think switching to init only would be, so we'd have to be careful about that one.

Otherwise, looks good, some minor comments here and there, but overall great!  Thanks for the changes!!

File: src/LanguageServer/Protocol/Protocol/IPartialResultParams.cs [hunk 1]
```diff
@@ -4,23 +4,23 @@
...
+        /// An <see cref="IProgress{T}"/> instance that can be used to report partial results
+        /// via the <c>$/progress</c> notification.
         /// </summary>
-        public IProgress<T>? PartialResultToken
-        {
-            get;
-            set;
-        }
+        [JsonPropertyName(Methods.PartialResultTokenName)]
+        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
```
[the remark this answers is not in the document]
Comment (mhutch) [reply src/LanguageServer/Protocol/Protocol/IPartialResultParams.cs:23]: The main reason I did this was to make it easy to copy/paste when implementing the interface 😄

File: src/LanguageServer/Protocol/Protocol/Internal/Diagnostics/VSInternalDocumentDiagnosticsParams.cs [hunk 2]
```diff
@@ -10,18 +10,14 @@ namespace Roslyn.LanguageServer.Protocol
...
+    internal class VSInternalDocumentDiagnosticsParams : VSInternalDiagnosticParams, IPartialResultParams<VSInternalDiagnosticReport[]>, IWorkDoneProgressParams
     {
-        /// <summary>
-        /// Gets or sets an optional token that a server can use to report work done progress.
-        /// </summary>
+        /// <inheritdoc/>
         [JsonPropertyName(Methods.WorkDoneTokenName)]
         [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
-        public IProgress<VSInternalDiagnosticReport[]>? WorkDoneToken { get; set; }
+        public IProgress<WorkDoneProgress> WorkDoneToken { get; set; }
```
[the remark this answers is not in the document]
Comment (mhutch) [reply src/LanguageServer/Protocol/Protocol/Internal/Diagnostics/VSInternalDocumentDiagnosticsParams.cs:18]: I'm pretty sure the previous WorkDoneProgress impl was plain wrong, so I'd be surprised if it was being used successfully

File: src/LanguageServer/Protocol/Protocol/WorkDoneProgress.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+/// An <c>IProgress&lt;WorkDoneProgress></c> is used to report progress to the server via the <c>$/progress</c> notification.
+/// <para>
+/// The derived classes <see cref="WorkDoneProgressBegin"/>, <see cref="WorkDoneProgressReport"/> and <see cref="WorkDoneProgressEnd"/>
+/// are used to report the beginning, progression, and end of the operation.
+/// </para>
+/// </summary>
+[JsonDerivedType(typeof(WorkDoneProgressBegin), "begin")]
+[JsonDerivedType(typeof(WorkDoneProgressReport), "report")]
+[JsonDerivedType(typeof(WorkDoneProgressEnd), "end")]
+[JsonPolymorphic(TypeDiscriminatorPropertyName = "kind")]
```
[the remark this answers is not in the document]
Comment (mhutch) [reply src/LanguageServer/Protocol/Protocol/WorkDoneProgress.cs:19]: Neither did I before this, so I hope it works 😄

File: src/LanguageServer/Protocol/Protocol/WorkDoneProgressBegin.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+/// <remarks>Since LSP 3.15</remarks>
+internal sealed class WorkDoneProgressBegin : WorkDoneProgress
+{
+    // NOTE: the kind property from the spec is used as a JsonPolymorphic discriminator on the base type
+
+    /// <summary>
+    /// Title of the progress operation. Used to briefly inform about the kind of operation being performed.
+    /// Examples: "Indexing" or "Linking dependencies".
+    /// </summary>
+    [JsonPropertyName("title")]
```
[the remark this answers is not in the document]
Comment (mhutch) [reply src/LanguageServer/Protocol/Protocol/WorkDoneProgressBegin.cs:24]: Yup I added a bunch of missing JsonRequired to existing members too but I may have missed a few

File: src/LanguageServer/Protocol/Protocol/Converters/InitializeParamsWorkspaceFoldersConverter.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+namespace Roslyn.LanguageServer.Protocol;
+
+// Per https://microsoft.github.io/language-server-protocol/specifications/specification-current/#initialize
+//
+// * If the property is NOT present, the client does not support workspace folders.
+// * If the property is present but null, the client supports workspace folders but none are open.
+//
+// To represent this, we use null to represent that the client does not support workspace
+// folders, and use an empty array to represent that none are open.
+internal class InitializeParamsWorkspaceFoldersConverter : JsonConverter<WorkspaceFolder[]?>
```
[the remark this answers is not in the document]
Comment (mhutch) [reply src/LanguageServer/Protocol/Protocol/Converters/InitializeParamsWorkspaceFoldersConverter.cs:17]: According to the spec, if the client sends a null value for the property then that means it supports workspace folders but none are open. If the client doesn't send the property at all, it doesn't support workspace folders. I couldn't think of another way to surface this distinction.

File: src/LanguageServer/Protocol/Protocol/Converters/InitializeParamsWorkspaceFoldersConverter.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+{
+    public override WorkspaceFolder[]? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+    {
+        return JsonSerializer.Deserialize<WorkspaceFolder[]>(ref reader, options) ?? [];
```
[the remark this answers is not in the document]
Comment (mhutch) [reply src/LanguageServer/Protocol/Protocol/Converters/InitializeParamsWorkspaceFoldersConverter.cs:21]: My intent here was that if the property is not present, we surface a value of null, and if the property is present but null, we surface an empty array, and if it's an empty array or non-empty array, we surface that array.

File: src/LanguageServer/Protocol/Protocol/SymbolInformation.cs [hunk 7]
```diff
@@ -10,9 +10,11 @@ namespace Roslyn.LanguageServer.Protocol
 
     /// <summary>
     /// Class representing information about programming constructs like variables, classes, interfaces, etc.
-    ///
+    /// <para>
     /// See the <see href="https://microsoft.github.io/language-server-protocol/specifications/specification-current/#symbolInformation">Language Server Protocol specification</see> for additional information.
+    /// </para>
     /// </summary>
+    [Obsolete("Use DocumentSymbol or WorkspaceSymbol instead")]
```
[the remark this answers is not in the document]
Comment (mhutch) [reply src/LanguageServer/Protocol/Protocol/SymbolInformation.cs:17]: Yes. This also causes obsoletion warnings elsewhere in the binding that will need to be suppressed.

File: src/LanguageServer/Protocol/Protocol/ServerCapabilities.cs [hunk 8]
```diff
@@ -81,28 +81,38 @@ internal class ServerCapabilities
...
 
         /// <summary>
-        /// Gets or sets a value indicating whether go to type definition is supported.
+        /// The server provides Go to Type Definition support.
         /// </summary>
+        /// <remarks>Since LSP 3.6</remarks>
         [JsonPropertyName("typeDefinitionProvider")]
         [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
-        public SumType<bool, TypeDefinitionOptions>? TypeDefinitionProvider { get; set; }
+        public SumType<bool, TypeDefinitionOptions, TypeDefinitionRegistrationOptions>? TypeDefinitionProvider { get; set; }
```
[the remark this answers is not in the document]
Comment (davidwengier) [reply src/LanguageServer/Protocol/Protocol/ServerCapabilities.cs:104]: Not yet, no

**Review by dibarbet — APPROVED** [review]
I think I'm good with this (as long as CI passes).  One small comment.  I didn't go through every file with a fine toothed comb though.

**Comment by mhutch** [comment]
@dibarbet i finally got it green!

**Pull request merged**